File: backend/shared/uidl/src/main/java/io/mateu/uidl/data/GridColumn.java
package io.mateu.uidl.data;

import lombok.Builder;

@Builder(toBuilder = true)
public record GridColumn(
    String id,
    String label,
    FieldDataType dataType,
    FieldStereotype stereotype,
    String style,
    String cssClasses,
    ColumnAlignment align,
    boolean sortable,
    String sortingProperty,
    boolean filterable,
    boolean frozen,
    boolean frozenToEnd,
    boolean autoWidth,
    String flexGrow,
    boolean resizable,
    String width,
    String tooltipPath,
    String actionId,
    String text,
    Integer priority,
    boolean identifier,
    boolean editable,
    Double weight)
    implements GridContent {

  public FieldDataType dataType() {
    return dataType != null ? dataType : FieldDataType.string;
  }

  public FieldStereotype stereotype() {
    return stereotype != null ? stereotype : FieldStereotype.regular;
  }

  /**
   * Returns the layout priority; lower value = higher importance. Defaults to Integer.MAX_VALUE.
   */
  public Integer priority() {
    return priority != null ? priority : Integer.MAX_VALUE;
  }
}

File: frontend/web/monorepo/libs/mateu/src/mateu/shared/apiClients/dtos/componentmetadata/GridColumn.ts
import ComponentMetadata from "@mateu/shared/apiClients/dtos/ComponentMetadata";

export default interface GridColumn extends ComponentMetadata {

    id: string
    label: string
    align: string
    sortable: boolean
    filterable: boolean
    frozen: boolean
    frozenToEnd: boolean
    autoWidth: boolean
    flexGrow: string
    resizable: boolean
    width: string
    dataType: string
    stereotype: string
    tooltipPath: string
    actionId: string
    text: string
    /** Rich `primary` column (coherence-plan #6): the row field for the secondary caption line,
     *  and the one for the leading avatar/icon. Set only when stereotype === 'primary'. */
    captionPath?: string
    leadingPath?: string
    style: string
    priority: number
    identifier: boolean
    editable: boolean
    editorType: string | undefined
    editorOptions: { value: any, label: string }[] | undefined
    weight: number | null
    /** Aggregate function totalled over the whole filtered set (and per group when grouping). */
    aggregate?: 'sum' | 'avg' | 'min' | 'max' | 'count'

}